builtin/providers/aws/resource_aws_dms_replication_instance.go

```diff
@@ -355,68 +399,35 @@ func resourceAwsDmsReplicationInstanceSetState(d *schema.ResourceData, instance
 	return nil
 }
 
-func waitForInstanceCreated(client *dms.DatabaseMigrationService, id string, delay int, maxAttempts int) error {
-	input := &dms.DescribeReplicationInstancesInput{
-		Filters: []*dms.Filter{
-			{
-				Name:   aws.String("replication-instance-id"),
-				Values: []*string{aws.String(id)},
-			},
-		},
-	}
+func resourceAwsDmsReplicationInstanceStateRefreshFunc(
+	d *schema.ResourceData, meta interface{}) resource.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+		conn := meta.(*AWSClient).dmsconn
 
-	config := waiter.Config{
-		Operation:   "DescribeReplicationInstances",
-		Delay:       delay,
-		MaxAttempts: maxAttempts,
-		Acceptors: []waiter.WaitAcceptor{
-			{
-				State:    "success",
-				Matcher:  "pathAll",
-				Argument: "ReplicationInstances[].ReplicationInstanceStatus",
-				Expected: "available",
+		v, err := conn.DescribeReplicationInstances(&dms.DescribeReplicationInstancesInput{
+			Filters: []*dms.Filter{
+				{
+					Name:   aws.String("replication-instance-id"),
+					Values: []*string{aws.String(d.Id())}, // Must use d.Id() to work with import.
+				},
 			},
-		},
-	}
-
-	w := waiter.Waiter{
-		Client: client,
-		Input:  input,
-		Config: config,
-	}
-
-	return w.Wait()
-}
+		})
+		if err != nil {
+			if dmserr, ok := err.(awserr.Error); ok && dmserr.Code() == "ResourceNotFoundFault" {
+				return nil, "", nil
+			}
+			log.Printf("Error on retrieving DMS Replication Instance when waiting: %s", err)
+			return nil, "", err
+		}
 
-func waitForInstanceDeleted(client *dms.DatabaseMigrationService, id string, delay int, maxAttempts int) error {
-	input := &dms.DescribeReplicationInstancesInput{
-		Filters: []*dms.Filter{
-			{
-				Name:   aws.String("replication-instance-id"),
-				Values: []*string{aws.String(id)},
-			},
-		},
-	}
+		if v == nil {
+			return nil, "", nil
+		}
 
-	config := waiter.Config{
-		Operation:   "DescribeReplicationInstances",
-		Delay:       delay,
-		MaxAttempts: maxAttempts,
-		Acceptors: []waiter.WaitAcceptor{
-			{
-				State:    "success",
-				Matcher:  "path",
-				Argument: "length(ReplicationInstances[]) > `0`",
-				Expected: false,
-			},
-		},
-	}
+		if v.ReplicationInstances != nil {
+			log.Printf("[DEBUG] DMS Replication Instance status for instance %s: %s", d.Id(), *v.ReplicationInstances[0].ReplicationInstanceStatus)
+		}
 
-	w := waiter.Waiter{
-		Client: client,
-		Input:  input,
-		Config: config,
+		return v, *v.ReplicationInstances[0].ReplicationInstanceStatus, nil
```

Comment:
This can panic if `v.ReplicationInstances` is `nil`, yeah? Not sure how likely that is, but seems worth guarding